### ISSUE TITLE
No double backwards implemented for several RNN ops

### ISSUE THREAD

**Issue by egrefen** [pytorch/pytorch#25315]
As flagged in pytorch/pytorch#5261, there are no second derivatives for several RNN modules when running on GPU due to CUDNN limitations. One proposed solution is to run the code for such modules with the `torch.backends.cudnn.flags(enabled=False)` context manager, but this yields runtime errors. Here is some repro code:

## This code block works
```python
model = torch.nn.Sequential(torch.nn.Linear(4, 3), torch.nn.Sigmoid()).to(gpu)
model_in = torch.rand(3, 4, requires_grad=True, device=gpu)
model_out = model(model_in)
model_first_grad = torch.autograd.grad(model_out.sum(), model_in, create_graph=True)[0]
model_second_grad = torch.autograd.grad(model_first_grad.sum(), model_in)[0]
print(model_second_grad) # this returns the second grad
```

## This code block returns `RuntimeError: derivative for _cudnn_rnn_backward is not implemented`.
```python
import torch
gpu = torch.device('cuda')

rnn_model = torch.nn.RNN(4, 4).to(gpu)
rnn_in = torch.rand(1, 3, 4, requires_grad=True, device=gpu)
rnn_out, _ = rnn_model(rnn_in)
rnn_first_grad = torch.autograd.grad(rnn_out.sum(), rnn_in, create_graph=True)[0]
rnn_second_grad = torch.autograd.grad(rnn_first_grad.sum(), rnn_in)[0]
print(rnn_second_grad) # this never runs
```

## This code block returns `RuntimeError: derivative for _thnn_fused_lstm_cell_backward is not implemented`.
```python
import torch
gpu = torch.device('cuda')

with torch.backends.cudnn.flags(enabled=False):
    rnn_model = torch.nn.LSTM(4, 4).to(gpu)
    rnn_in = torch.rand(1, 3, 4, requires_grad=True, device=gpu)
    rnn_out, _ = rnn_model(rnn_in)
    rnn_first_grad = torch.autograd.grad(rnn_out.sum(), rnn_in, create_graph=True)[0]
    rnn_second_grad = torch.autograd.grad(rnn_first_grad.sum(), rnn_in)[0]
    print(rnn_second_grad) # this never runs
```
    
## This code block returns `RuntimeError: derivative for _thnn_fused_gru_cell_backward is not implemented`.
```python
import torch
gpu = torch.device('cuda')

with torch.backends.cudnn.flags(enabled=False):
    rnn_model = torch.nn.GRU(4, 4).to(gpu)
    rnn_in = torch.rand(1, 3, 4, requires_grad=True, device=gpu)
    rnn_out, _ = rnn_model(rnn_in)
    rnn_out = rnn_out.pow(2)  # to ensure hessian exists
    rnn_first_grad = torch.autograd.grad(rnn_out.sum(), rnn_in, create_graph=True)[0]
    rnn_second_grad = torch.autograd.grad(rnn_first_grad.sum(), rnn_in)[0]
    print(rnn_second_grad) # this never runs
```

Could the second derivative code for these three types of RNNs (and others in the codebase) be implemented? This would be a requirement to be able to run meta-learning algorithms like MAML in pytorch on GPU for RNNs.

cc @ezyang @SsnL @albanD @zou3519

**Comment by egrefen** [comment]
> It's near impossible to implement double backwards for cudnn RNN ops because a lot of the necessary information is hidden.

Sorry if I wasn't clear, but we're specifically just looking for non-CUDNN double backward support on GPU.

> It's possible to write a double backwards for our fused implementation. However, I would point users at writing a TorchScript RNN ([tutorial here](https://pytorch.org/blog/optimizing-cuda-rnn-with-torchscript/)) instead because the JIT is able to generate fused code that is close in performance to our fused implementations.

This has been pointed out, but the issue is that what are are trying to do has incompatibility issues with torchscript which are being separately addressed.

**Comment by soumith** [comment]
i can take a crack at double backward for `_thnn_fused_lstm_cell_backward` on a long flight. will report back what I find

**Comment by ngimel** [comment]
Doing autograddable backward that would allow automatic double backward is probably easier, similar to how it is done for weight norm now.

**Comment by egrefen** [comment]
Any updates? Was it solvable on one transatlantic flight, @soumith? 😄

**Comment by ngimel** [comment]
Fixed in #26660
